core/test/transform_test.cpp

```diff
@@ -902,3 +902,39 @@ TEST_F(TransformInPlaceDryRunTest, unchanged_if_success) {
   dry_run::transform_in_place<pair_self_t<double>>(a, a, binary);
   EXPECT_EQ(a, original);
 }
+
+TEST(TransformFlagsTest, variance_on_arg_throws) {
+  auto a = makeVariable<double>(Values{1}, Variances{1});
+  auto b = makeVariable<double>(Values{1});
+  auto binary_op = [](auto x, auto y) { return x + y; };
+  auto op_arg_0_has_flags =
+      scipp::overloaded{transform_flags::expect_no_variance_arg<0>, binary_op};
+  EXPECT_THROW(transform<std::tuple<double>>(a, b, op_arg_0_has_flags),
+               except::VariancesError);
+  EXPECT_NO_THROW(transform<std::tuple<double>>(b, a, op_arg_0_has_flags));
+  EXPECT_NO_THROW(transform<std::tuple<double>>(b, b, op_arg_0_has_flags));
+  auto op_arg_1_has_flags =
+      scipp::overloaded{transform_flags::expect_no_variance_arg<1>, binary_op};
+  EXPECT_THROW(transform<std::tuple<double>>(b, a, op_arg_1_has_flags),
+               except::VariancesError);
+  EXPECT_NO_THROW(transform<std::tuple<double>>(a, b, op_arg_1_has_flags));
+  EXPECT_NO_THROW(transform<std::tuple<double>>(b, b, op_arg_1_has_flags));
```

Comment:
Also test with the flag for both args? The code in transform is intricate enough to fear that that behaves differently.